### PR TITLE
Generate text with DALLE

### PR DESCRIPTION
**Pull request by jules-samaran**
Since DALLE trains a multimodal language model, the text part of the sequence can also be generated from scratch.
I added a new method to generate text in the DALLE class and also an argument in generate.py so that the generated image can be conditioned on a generated text instead of an input text.
To make this work I had to add an "ignore_tokens" argument to the decoder method of each argument otherwise the tokenizers raise an error when trying to decode padding tokens.

### CODE REVIEW

File: dalle_pytorch/tokenizer.py [hunk 1]
```diff
@@ -164,11 +164,11 @@ def __init__(self, bpe_path = None):
         self.tokenizer = tokenizer
         self.vocab_size = tokenizer.get_vocab_size()
 
-    def decode(self, tokens):
+    def decode(self, tokens, ignore_pad_tokens = []):
```
Comment (rom1504) [inline dalle_pytorch/tokenizer.py:167]: the variable used below seems to be pad_tokens not ignore_pad_tokens

File: dalle_pytorch/tokenizer.py [hunk 2]
```diff
@@ -164,11 +164,11 @@ def __init__(self, bpe_path = None):
...
         if torch.is_tensor(tokens):
             tokens = tokens.tolist()
-
-        tokens = [token for token in tokens if token not in (0,)]
+        ignore_ids = [0] + pad_tokens
+        tokens = [token for token in tokens if token not in ignore_ids]
```
Comment (rom1504) [inline dalle_pytorch/tokenizer.py:171]: what do you think about making this a set instead of a list ? the `in` operator of a set is O(1) instead of O(n) for a list. It will be a little bit faster

File: dalle_pytorch/dalle_pytorch.py [hunk 3]
```diff
@@ -528,3 +528,51 @@ def forward(
 
         loss = (loss_text + self.loss_img_weight * loss_img) / (self.loss_img_weight + 1)
         return loss
+    
+    
+    
+    @torch.no_grad()
+    @eval_decorator
+    def generate_text(
+        self,
```
Comment (rom1504) [inline dalle_pytorch/dalle_pytorch.py:537]: being able to provide a text here could be interesting.
Instead of starting from scratch, that could make it possible to complete an initial text

**Comment by rom1504** [comment]
Overall it looks pretty good :), I left a few comments

File: dalle_pytorch/tokenizer.py (hunk 1 above)
Comment (jules-samaran) [reply dalle_pytorch/tokenizer.py:167]: Nice catch

File: dalle_pytorch/tokenizer.py (hunk 2 above)
Comment (jules-samaran) [reply dalle_pytorch/tokenizer.py:171]: Good idea, I've just fixed that

**Comment by rom1504** [comment]
I cleaned up a few things and pushed in https://github.com/lucidrains/DALLE-pytorch/pull/332 ; can you tell me if that looks good to you @jules-samaran ?

**Comment by rom1504** [comment]
I cleaned up a few things and tested it. It works for all cases
* generating images from a text
* generating image+text from nothing
* completing text from a text and generating images from that

example of use put in the readme
```python
python generate.py --dalle_path ./dalle.pt --text 'a dog chewing a bone' --gentext
```

**Comment by jules-samaran** [comment]
Looks perfect to me!

**Comment by rom1504** [comment]
One example of result with empty text as input : 
> puma adults - men men black fashion fall 2 0 1 1 casual casual shoes footwear shoes 
image

I think this can not only be used to generate some interesting results, but also to understand a bit more what's the consequence of using a particular tokenizer (for example this simple one is separating all the digits of numbers)

**Comment by rom1504** [comment]
ready to review/merge @lucidrains

**Review by rom1504 — APPROVED** [review]
